diff --git a/lib/Andersens/Andersens.cpp b/lib/Andersens/Andersens.cpp
index 4f86fb1..d2650c8 100644
--- a/lib/Andersens/Andersens.cpp
+++ b/lib/Andersens/Andersens.cpp
@@ -699,8 +699,10 @@ AliasAnalysis::AliasResult Andersens::alias(const Location &L1,
 
   // Check to see if the two pointers are known to not alias.  They don't alias
   // if their points-to sets do not intersect.
-  if (!N1->intersectsIgnoring(N2, NullObject))
-    return NoAlias;
+  if (!N1->PointsTo->test(UniversalSet) && !N2->PointsTo->test(UniversalSet)) {
+    if (!N1->intersectsIgnoring(N2, NullObject))
+      return NoAlias;
+  }
 
   return AliasAnalysis::alias(L1, L2);
 }
